plugins/quickicon/eos/src/Extension/Eos.php

```diff
@@ -0,0 +1,347 @@
+<?php
+
+/**
+ * @package     Joomla.Plugin
+ * @subpackage  Quickicon.eos
+ *
+ * @copyright   (C) 2023 Open Source Matters, Inc. <https://www.joomla.org>
+ * @license     GNU General Public License version 2 or later; see LICENSE.txt
+ */
+
+namespace Joomla\Plugin\Quickicon\Eos\Extension;
+
+use Exception;
+use Joomla\CMS\Access\Exception\NotAllowed;
+use Joomla\CMS\Cache\CacheController;
+use Joomla\CMS\Factory;
+use Joomla\CMS\HTML\HTMLHelper;
+use Joomla\CMS\Language\Text;
+use Joomla\CMS\Plugin\CMSPlugin;
+use Joomla\Database\DatabaseAwareTrait;
+use Joomla\Event\SubscriberInterface;
+use Joomla\Module\Quickicon\Administrator\Event\QuickIconsEvent;
+
+// phpcs:disable PSR1.Files.SideEffects
+\defined('_JEXEC') or die;
+// phpcs:enable PSR1.Files.SideEffects
+
+/**
+ * Joomla! end of support notification plugin
+ *
+ * @since __DEPLOY_VERSION__
+ */
+final class Eos extends CMSPlugin implements SubscriberInterface
+{
+    use DatabaseAwareTrait;
+
+    /**
+     * The EOS date for 4.4. and beyond
+     *
+     * @var    string
+     * @since __DEPLOY_VERSION__
+     */
+    public const EOS_DATE = '2023-10-25';
+
+    /**
+     * Load the language file on instantiation.
+     *
+     * @var    bool
+     * @since  __DEPLOY_VERSION__
+     */
+    protected $autoloadLanguage = true;
+
+    /**
+     * Holding the current valid message to be shown
+     *
+     * @var    array
+     * @since __DEPLOY_VERSION__
+     */
+    private array $currentMessage = [];
+
+    /**
+     * Are the messages initialised
+     *
+     * @var    bool
+     * @since __DEPLOY_VERSION__
+     */
+
+    private bool $messagesInitialized = false;
+
+    /**
+     * Returns an array of events this subscriber will listen to.
+     *
+     * @return  array
+     *
+     * @since __DEPLOY_VERSION__
+     */
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            'onGetIcons' => 'getEndOfServiceNotification',
+            'onAjaxEos'  => 'onAjaxEos',
+        ];
+    }
+
+    /**
+     * Check and show the the alert and quickicon message
+     *
+     * This method is called when the Quick Icons module is constructing its set
+     * of icons. You can return an array which defines a single icon and it will
+     * be rendered right after the stock Quick Icons.
+     *
+     * @param   QuickIconsEvent  $event  The event object
+     *
+     * @return  void
+     *
+     * @since __DEPLOY_VERSION__
+     *
+     * @throws Exception
+     */
+    public function getEndOfServiceNotification(QuickIconsEvent $event): void
+    {
+        if ($event->getContext() !== $this->params->get('context', 'update_quickicon') || !$this->shouldDisplayMessage() || !$this->messagesInitialized && $this::setMessage() == []) {
```

Comment:
```suggestion
        if ($event->getContext() !== $this->params->get('context', 'update_quickicon') || !$this->shouldDisplayMessage() || !$this->messagesInitialized && $this->setMessage() == []) {
```
why use _Scope Resolution Operator_?

Reply:
Because phpstorm says it's not defined except as a :: function

Reply:
But in line 287 below you use `$this->setMessage()`. Shouldn't phpstorm have complained there, too? I think `$this->setMessage()` should work here.